Source/Tools/FEXLoader/LinuxSyscalls/SignalDelegator.cpp

```diff
@@ -1376,13 +1383,105 @@ namespace FEX::HLE {
 
   /**  @} */
 
+  static bool IsAsyncSignal(siginfo_t* Info, int Signal) {
+    if (Info->si_code <= SI_USER) {
+      // If the signal is not from the kernel then it is always async.
+      // This is because synchronous signals can be sent through tgkill,sigqueue and other methods.
+      // SI_USER == 0 and all negative si_code values come from the user.
+      return true;
+    }
+    else {
+      // If the signal is from the kernel then it is async only if it isn't an explicit synchronous signal.
+      switch (Signal) {
+        // These are all synchronous signals.
+        case SIGBUS:
+        case SIGFPE:
+        case SIGILL:
+        case SIGSEGV:
+        case SIGTRAP:
+          return false;
+        default: break;
+      }
+    }
+
+    // Everything else is async and can be deferred.
+    return true;
+  }
+
   void SignalDelegator::HandleGuestSignal(FEXCore::Core::InternalThreadState *Thread, int Signal, void *Info, void *UContext) {
+    ucontext_t* _context = (ucontext_t*)UContext;
+    auto SigInfo = *static_cast<siginfo_t*>(Info);
+
+    constexpr bool SupportDeferredSignals = true;
+    if (SupportDeferredSignals) {
+      auto RefCount = Thread->CurrentFrame->State.DeferredSignalRefCount.Load();
+
+      if (Signal == SIGSEGV &&
+          SigInfo.si_code == SEGV_ACCERR &&
+          SigInfo.si_addr == reinterpret_cast<void*>(Thread->CurrentFrame->State.DeferredSignalFaultAddress)) {
+        if (RefCount == 0) {
+          // We are faulting with an attempt to check deferred signals.
+          // Pull a signal frame off the stack.
+
+          mprotect(reinterpret_cast<void*>(Thread->CurrentFrame->State.DeferredSignalFaultAddress), 4096, PROT_READ | PROT_WRITE);
+
+          if (Thread->DeferredSignalFrames.size() == 0) {
+            // No signals to defer. Just set the fault page back to RW and continue execution.
+            // This occurs as a minor race condition between the refcount decrement and the access to the fault page.
+            return;
+          }
+          else {
+            auto Top = Thread->DeferredSignalFrames.back();
+            Signal = Top.Signal;
+            SigInfo = Top.Info;
+            Thread->DeferredSignalFrames.pop_back();
+
+            // Now that we are handling a deferred signal state. mprotect the fault page with RW permissions.
+            // This puts FEX in a state in-which we are *permanently* deferring signals and /not/ checking them.
```

Comment:
This was a bit confusing to read since we've already mprotect'ed at this point. This is more succinct:

```suggestion
            // Until we re-protect the page to PROT_NONE, FEX will now *permanently* defer signals and /not/ check them.
```

Reply:
Resolved.